src/bin/nydus-image/merge.rs

```diff
@@ -0,0 +1,183 @@
+            } else {
+                // Keep final bootstrap following superblock flags of source bootstraps.
+                current_flags.set_to_ctx(ctx);
+                flags = Some(current_flags);
+            }
+
+            let parent_blobs = rs.superblock.get_blob_infos();
+            let blob_hash = Self::get_blob_hash(bootstrap_path)?;
+            let mut blob_idx_map = Vec::new();
+            let mut parent_blob_added = false;
+
+            for blob in &parent_blobs {
+                let mut blob_ctx = BlobContext::from(blob, ChunkSource::Parent);
+                if chunk_dict_blobs.get(blob.blob_id()).is_none() {
+                    // Only up to one blob from the parent bootstrap, the other blobs should be
+                    // from the chunk dict image.
+                    if parent_blob_added {
+                        bail!("invalid bootstrap, seems have multiple non-chunk-dict blobs in this bootstrap");
+                    }
+                    // The blob id (blob sha256 hash) in parent bootstrap is invalid for nydusd
+                    // runtime, should change it to the hash of whole tar blob.
+                    blob_ctx.blob_id = blob_hash.to_owned();
+                    parent_blob_added = true;
+                }
+                blob_idx_map.push(blob_mgr.len() as u32);
+                blob_mgr.add(blob_ctx);
+            }
+
+            if let Some(tree) = &mut tree {
+                let mut nodes = Vec::new();
+                rs.walk_inodes(RAFS_ROOT_INODE, None, &mut |inode: &dyn RafsInode,
+                                                            path: &Path|
+                 -> Result<()> {
+                    let mut node = MetadataTreeBuilder::parse_node(&rs, inode, path.to_path_buf())
+                        .context(format!("parse node from bootstrap {:?}", bootstrap_path))?;
+                    for chunk in &mut node.chunks {
+                        let origin_blob_index = chunk.inner.blob_index() as usize;
+                        // Set the blob index of chunk to real index in blob table of final bootstrap.
+                        chunk.inner.set_blob_index(blob_idx_map[origin_blob_index]);
+                    }
+                    // Set node's layer index to distinguish same inode number (from bootstrap)
+                    // between different layers.
+                    node.layer_idx = u16::try_from(layer_idx).context(format!(
+                        "too many layers {}, limited to {}",
+                        layer_idx,
+                        u16::MAX
+                    ))?;
+                    node.overlay = Overlay::UpperAddition;
+                    match node.whiteout_type(WhiteoutSpec::Oci) {
+                        Some(_) => {
+                            // Insert removal operations at the head, so they will be handled first when
+                            // applying to lower layer.
+                            nodes.insert(0, node);
```

Comment:
nits: `Insert whiteouts at the head, xxxx`

Reply:
Fixed